### DISCUSSION TITLE
How to see my git secrets?

### DISCUSSION THREAD

**Discussion by Fl4m3Ph03n1x** [community/community#26277] [Actions]
<h2>Background</h2>
<p>I had some secrets in my code and upon learning about GitHub Actions I decided to save them in the repositories secret menu for later use in my pipeline.</p>
<h2>Problem</h2>
<p>However, now I need to access this secrets to develop a new feature and I can’t. Every time I try to see the value it asks me to update the secrets. I dont want to update anything I just want to see their values.</p>
<h2>Question</h2>
<p>How can i see the unencrypted values of my secrets in the project?</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/Fl4m3Ph03n1x">@Fl4m3Ph03n1x</a>,</p>
<p>I recommend that you should avoid trying to output the values of repository secrets in the workflow runs. This may lead to the risk that the secrets are leaked.<br>
When you try to print secrets to the log, GitHub automatically redacts secrets printed to the log, the values of the secrets will be masked and displayed as “<strong>***</strong>”.</p>
<p>If you want to use the repository secrets in the source code files of your project when you build the project in the workflow runs, maybe you can try the following methods.</p>
<ol>
<li>If the source code files can recognize and access the system environment variables, you can set the secrets as environment variables on the runner machine.<br>
In the workflow files, you can use the <a href="https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#env" rel="nofollow noopener">'<strong>env</strong></a>’ key to set the environment variables.</li>
</ol>
<pre><code class="lang-yaml">env:
  ENV_VAR_NAME: ${{ secrets.SECRET_NAME }}
</code></pre>
<p>In the source code files, use the appropriate expressions to access environment variables. The expressions may be different between different types of the files. The expressions may include “<strong>${ENV_VAR_NAME}</strong>”, “<strong>$ENV_VAR_NAME</strong>”, “<strong>$env:ENV_VAR_NAME</strong>”, etc…</p>
<ol start="2">
<li>If the source code files can’t recognize the mapped system environment variables, you can use the command (such as ‘<strong>sed</strong>’) to directly replace the expressions in the source code files with the value of the environment variables.<br>
For example:</li>
</ol>
<pre><code class="lang-bash">sed -i 's/${PKG_VERSION}/${{ env.PKG_VERSION }}/g' package.json
</code></pre>
<p>Below is a simple example that is applying both of the above two methods.<br>
</p><aside class="onebox githubblob">
  <header class="source">
      <a href="https://github.com/TestGitHubPackagesRML/TestNpm/blob/080990526adbb61b0365e4b207d972de7d46dc77/.npmrc#L1" target="_blank" rel="nofollow noopener">github.com</a>
  </header>
  <article class="onebox-body">
    <h4><a href="https://github.com/TestGitHubPackagesRML/TestNpm/blob/080990526adbb61b0365e4b207d972de7d46dc77/.npmrc#L1" target="_blank" rel="nofollow noopener">TestGitHubPackagesRML/TestNpm/blob/080990526adbb61b0365e4b207d972de7d46dc77/.npmrc#L1</a></h4>
<pre class="onebox"><code class="lang-npmrc"><ol class="start lines" start="1" style="counter-reset: li-counter 0 ;">
<li class="selected">//npm.pkg.github.com/:_authToken=${GITHUB_TOKEN}</li>
<li>registry=https://npm.pkg.github.com/testgithubpackagesrml</li>
</ol></code></pre>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>
<br>
<aside class="onebox githubblob">
  <header class="source">
      <a href="https://github.com/TestGitHubPackagesRML/TestNpm/blob/080990526adbb61b0365e4b207d972de7d46dc77/package.json#L3" target="_blank" rel="nofollow noopener">github.com</a>
  </header>
  <article class="onebox-body">
    <h4><a href="https://github.com/TestGitHubPackagesRML/TestNpm/blob/080990526adbb61b0365e4b207d972de7d46dc77/package.json#L3" target="_blank" rel="nofollow noopener">TestGitHubPackagesRML/TestNpm/blob/080990526adbb61b0365e4b207d972de7d46dc77/package.json#L3</a></h4>
<pre class="onebox"><code class="lang-json"><ol class="start lines" start="1" style="counter-reset: li-counter 0 ;">
<li>{</li>
<li>  "name": "@TestGitHubPackagesRML/testnpm",</li>
<li class="selected">  "version": "${PKG_VERSION}",</li>
<li>  "description": "This package is used to test npm for use with GitHub Packages.",</li>
<li>  "main": "index.js",</li>
<li>  "scripts": {</li>
<li>    "test": "echo \"Error: no test specified\" &amp;&amp; exit 1"</li>
<li>  },</li>
<li>  "repository": {</li>
<li>    "type": "git",</li>
<li>    "url": "git+https://github.com/TestGitHubPackagesRML/TestNpm.git"</li>
<li>  },</li>
<li>  "author": "",</li>
</ol></code></pre>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>
<br>
<aside class="onebox githubblob">
  <header class="source">
      <a href="https://github.com/TestGitHubPackagesRML/TestNpm/blob/080990526adbb61b0365e4b207d972de7d46dc77/.github/workflows/CI.yml#L27" target="_blank" rel="nofollow noopener">github.com</a>
  </header>
  <article class="onebox-body">
    <h4><a href="https://github.com/TestGitHubPackagesRML/TestNpm/blob/080990526adbb61b0365e4b207d972de7d46dc77/.github/workflows/CI.yml#L27" target="_blank" rel="nofollow noopener">TestGitHubPackagesRML/TestNpm/blob/080990526adbb61b0365e4b207d972de7d46dc77/.github/workflows/CI.yml#L27</a></h4>
<pre class="onebox"><code class="lang-yml"><ol class="start lines" start="17" style="counter-reset: li-counter 16 ;">
<li>  uses: actions/checkout@v2</li>
<li>
</li>
<li>- name: Setup node</li>
<li>  uses: actions/setup-node@v1</li>
<li>  with:</li>
<li>    node-version: 12</li>
<li>
</li>
<li>- name: Set package version</li>
<li>  run: echo "::set-env name=PKG_VERSION::${{ env.MAIN_VER }}.${{ github.run_number }}.$(date +"%y%m%d")"</li>
<li>
</li>
<li class="selected">- name: Publish package</li>
<li>  run: |</li>
<li>    sed -i 's/${PKG_VERSION}/${{ env.PKG_VERSION }}/g' package.json</li>
<li>    npm publish</li>
<li>  env:</li>
<li>    GITHUB_TOKEN: ${{secrets.GITHUB_TOKEN}}</li>
</ol></code></pre>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>
<p></p>

**Comment by Fl4m3Ph03n1x** [comment]
<p>Thank you for your response.</p>
<p>My app can read the secrets from environment variables and is indeed doing so. My question is more aimed at “How can I, a  human being, see the secrets I saved earlier?” Is this even possible?</p>
<p>Let’s say I need to show someone the content of these secrets, how can I do it?</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/Fl4m3Ph03n1x">@Fl4m3Ph03n1x</a>,</p>
<p>We have no safe way to fetch the values of secrets, and actually it is not recommended to try fetching and sharing the secrets.</p>

**Reply by kostyanius** [reply]
the most dumb solution for ci/cd that i have ever faced.

**Comment by mirdhyn** [comment]
<p>while there is no safe way to display the secret, the tricky &amp; unsafe way (if you really need to) is to run something like <code>echo ${{secrets.SECRET_NAME}} | sed 's/./&amp; /g'</code> in a workflow, that would output the secret with spaces between each character, Github will not recognize the secret and will not hide it.<br>
Beware that the secret will then be in the actions logs in clear form, and everybody having access to the repo will be able to see it.</p>

**Reply by rocketteering** [reply]
This worked perfectly, thanks!

**Comment by lukepon** [comment]
<p>Is there any way to avoid this? For instance if you have a serviceaccount.json in your gitaction secret and a potentially nefarious actor wants to read it? It seems that this is pretty damning of gitactions because the secrets are scoped to people who have access to the repository and not people themselves (or is that wrong?)</p>

**Comment by airtower-luna** [comment]
<p>Then underlying assumption is basically that anyone with <em>direct</em> write access to the repository is trusted. After all, with that access they could put malicious things directly into the code.</p>
<p>Forks are a different story, see <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/encrypted-secrets#using-encrypted-secrets-in-a-workflow">Using encrypted secrets in a workflow</a>:</p>
<blockquote>
<p>With the exception of <code>GITHUB_TOKEN</code> , secrets are not passed to the runner when a workflow is triggered from a forked repository.</p>
</blockquote>
<p>So someone making a pull request would have to trick you into merging the secret-stealing code.</p>

**Comment by lukepon** [comment]
<p>You make a good point about the write access. But injecting malicious code seems like a whole different problem, dependencies alone may cause this.<br>
My concern is more that after setting up a pipeline for a team that somebody may think that those aws secrets or whatever; would be useful to have locally for testing or for giving to others so they can quickly change something (like injecting malicious code).<br>
What I’ve set up now is basically an audit log to see who is accessing resources with which credentials  but it would save me and mine time if the secrets could be scoped to certain teams or individuals, in line with the principle of least privilige, instead of a repository.</p>

**Comment by scottmelhop** [comment]
<p>Github secrets are not really that secret, especially when combined with Github actions. You can for example have your github action create a new branch, create a new file, write your secrets, add it to the new branch, commit it and push it. Then you can navigate to that branch in the UI and the secret will be there for everyone to see. So as long as Github actions are enabled, no secret is safe.</p>

**Comment by unfor19** [comment]
<p><a class="mention" href="/fl4m3ph03n1x">@Fl4m3Ph03n1x</a></p>
<p>To answer your question:</p>
<blockquote>
<p>Let’s say I need to show someone the content of these secrets, how can I do it?</p>
</blockquote>
<pre><code class="lang-yaml">name: Recovering secrets

# Assumption:
# You've created the following GitHub secrets in your repository:
# MY_CLIENT_SECRET - encrypt/decrypt with openssl - useful for public and public repositories
# MY_OPENSSL_PASSWORD - used to protect secrets
# MY_OPENSSL_ITER - Use a number of iterations on the password to derive the encryption key.
#                   High values increase the time required to brute-force the resulting file.
#                   This option enables the use of PBKDF2 algorithm to derive the key.

on:
  push:
  workflow_dispatch:

jobs:
  openssl:
    name: Recover With OpenSSL
    runs-on: ubuntu-20.04
    steps:
      - uses: actions/checkout@v3
      - env:
          MY_CLIENT_SECRET: ${{ secrets.MY_CLIENT_SECRET }}
          MY_OPENSSL_PASSWORD: ${{ secrets.MY_OPENSSL_PASSWORD }}
          MY_OPENSSL_ITER: ${{ secrets.MY_OPENSSL_ITER }}
        run: |
          echo "MY_CLIENT_SECRET (***)     = ${MY_CLIENT_SECRET}"
          echo "MY_CLIENT_SECRET (openssl) = $(echo "${MY_CLIENT_SECRET}" | openssl enc -e -aes-256-cbc -a -pbkdf2 -iter ${MY_OPENSSL_ITER} -k "${MY_OPENSSL_PASSWORD}")"
          echo "Copy the above value, and then execute locally:"
          echo "echo PASTE_HERE | openssl base64 -d | openssl enc -d -pbkdf2 -iter \$MY_OPENSSL_ITER -aes-256-cbc -k \$MY_OPENSSL_PASSWORD"
</code></pre>
<p>The trick is to encrypt the secret before printing it, with a known <code>MY_OPENSSL_ITER</code> and <code>MY_OPENSSL_PASSWORD</code>. That way, you can locally decrypt it; I even wrote a blog post about this specific topic as it really bothered me that it’s not possible to recover secrets once they’re set - <a href="https://dev.to/unfor19/how-to-recover-secrets-from-github-actions-5k2" class="inline-onebox" rel="noopener nofollow ugc">How To Recover Secrets From GitHub Actions - DEV Community</a></p>

**Comment by ClaireCJS** [comment]
It would be nice if there were some sort of comment we could insert

Much like the comments to tell pylint to stop nagging about thing sin python   "# pylint: disable=W0103"

If we could put a comment like    "# github: secret"

which would then cause github to replace it _in the checked out code, that would be fantastic too.

**Reply by airtower-luna** [reply]
Technically, you could use something similar to

```yaml
run: sed -i "/@MY_SECRET@/$MY_SECRET/g" my_file.py
env:
  MY_SECRET: ${{ secrets.MY_SECRET }}
```

to search and replace any occurrence of `@MY_SECRET@` in the source file.

However, I'd strongly advise against it, and against anything else that requires secrets in source files. It's too easy to make mistakes that way that reveal secrets, and annoyingly hard to change them when necessary. :warning: 

A better approach is to read secrets from a configuration file or environment variables, whatever works better for you. That way your workflow can just write/set that, and call your code without any modification.

**Reply by ClaireCJS** [reply]
Oh I totally agree, but i still believe this should be a feature. It doesn't hurt to have extra layers of protection. People make mistakes.

**Reply by airtower-luna** [reply]
I might have misunderstood what you want then. I thought you were looking for a way to insert secrets into your code files, which would *remove* a layer of protection. Did you mean something else? :thinking:

**Reply by ClaireCJS** [reply]
I was perfectly clear, twice. I'm not repeating myself.

**Reply by tatyree** [reply]
A bit late to the party, but my solution to this problem was to set up a Firebase Cloud Function on an existing, experimental project that already had Could Functions and Firestore configured.  In one of my GH Action configurations, I used [HTTP Request Action](https://github.com/marketplace/actions/http-request-action) from the GH Actions Marketplace to post each secret to the endpoint.  I configured the function to add a new record to a temporary Firestore collection with each post.  Committing the HTTP Request Action   configuration was enough to get a complete record of all of my secrets.  Finally, I removed the HTTP Request Action configuration from my next commit, copied the secrets out of Firestore using the web console and deleted the Firestore collection.

**Comment by gajus** [comment]
Just do:

```yaml
jobs:
  debug_secrets:
    name: 'debug secrets'
    steps:
        run: echo '${{ toJSON(secrets) }}' | base64
name: '[apps] debug secrets'
on:
  pull_request:
    types:
      - opened
      - synchronize
      - reopened
```

**Comment by dennisvang** [comment]
related: https://github.com/cli/cli/discussions/3397

**Comment by allohamora** [comment]
related: https://github.com/orgs/community/discussions/26277
